File: copilot-demo-labs/use-cases/10-prompt-engineering-advanced/src/order_management.js

const TAX_RATE = 0.22;

class Inventory {
  constructor() {
    this.stock = {};
  }

  restock(product, qty) {
    this.stock[product] = (this.stock[product] ?? 0) + qty;
  }

  check(product, qty) {
    return (this.stock[product] ?? 0) >= qty;
  }

  reserve(product, qty) {
    this.stock[product] -= qty;
  }
}

class Notification {
  sendConfirmation(orderId, total, items) {
    console.log(`Order ${orderId} confirmed: ${JSON.stringify(items)}, total=${total}`);
  }
}

class Order {
  constructor(inventory, notification) {
    this.inventory = inventory;
    this.notification = notification;
    this.counter = 0;
  }

  place(items, discountPct = 0) {
    for (const item of items) {
      if (!this.inventory.check(item.product, item.qty)) {
        throw new Error(`Out of stock: ${item.product}`);
      }
    }

    const subtotal = items.reduce((sum, i) => sum + i.price * i.qty, 0);

    // BUG: tax applied before discount
    const taxed = subtotal + subtotal * TAX_RATE;
    const total = taxed - (taxed * discountPct) / 100;

    for (const item of items) {
      this.inventory.reserve(item.product, item.qty);
    }

    this.counter++;
    const orderId = `ORD-${String(this.counter).padStart(4, '0')}`;

    // BUG: notification never called

    return { orderId, total: Math.round(total * 100) / 100, items };
  }
}

module.exports = { Inventory, Notification, Order };

File: copilot-demo-labs/use-cases/02-prompt-engineering-advanced/src/OrderManagement.cs
namespace Demo;

public record OrderItem(string Product, int Qty, decimal Price);

public class Inventory
{
    private readonly Dictionary<string, int> _stock = new();

    public void Restock(string product, int qty)
    {
        _stock[product] = _stock.GetValueOrDefault(product) + qty;
    }

    public bool Check(string product, int qty)
    {
        return _stock.GetValueOrDefault(product) >= qty;
    }

    public void Reserve(string product, int qty)
    {
        _stock[product] -= qty;
    }
}

public class Notification
{
    public void SendConfirmation(string orderId, decimal total, List<OrderItem> items)
    {
        Console.WriteLine($"Order {orderId} confirmed: {string.Join(", ", items)}, total={total}");
    }
}

public class Order
{
    private const decimal TaxRate = 0.22m;
    private readonly Inventory _inventory;
    private readonly Notification _notification;
    private int _counter;

    public Order(Inventory inventory, Notification notification)
    {
        _inventory = inventory;
        _notification = notification;
    }

    public (string OrderId, decimal Total, List<OrderItem> Items) Place(List<OrderItem> items, decimal discountPct = 0)
    {
        foreach (var item in items)
        {
            if (!_inventory.Check(item.Product, item.Qty))
                throw new InvalidOperationException($"Out of stock: {item.Product}");
        }

        var subtotal = items.Sum(i => i.Price * i.Qty);

        // BUG: tax applied before discount
        var taxed = subtotal + subtotal * TaxRate;
        var total = taxed - taxed * discountPct / 100;

        foreach (var item in items)
        {
            _inventory.Reserve(item.Product, item.Qty);
        }

        _counter++;
        var orderId = $"ORD-{_counter:D4}";

        // BUG: notification never called

        return (orderId, Math.Round(total, 2), items);
    }
}
